test/render.test.jsx

```diff
@@ -171,80 +166,78 @@ test('useEffect(f, [x]) should run on changes to x', async () => {
 
   await testUpdates([
     {
-      content: <Component value={0} />,
+      content: <Component value={1} />,
       test: () => {
-        nextTick()
-        expect(effects).toEqual([])
+        expect(effects).toEqual(['effect 1'])
+        effects = []
       }
     },
     {
-      content: <Component value={1} />,
+      content: <Component value={2} />,
       test: () => {
-        nextTick()
-        expect(effects).toEqual(['effect 0'])
+        expect(effects).toEqual(['cleanUp 1', 'effect 2'])
+        effects = []
       }
     },
     {
-      content: <Component value={1} />,
+      content: <Component value={2} />,
```

Comment:
because the new version of fre has built-in shouldComponentUpdate by default. When props are equal, the child component will not be rerender, so make the props to be different.
```js
content: <Component value={2} count={0}/> // add any props to rerender
```

Reply:
The point of this step of the test is to show what happens when effect dependencies *don't* change - the test is already passing, but maybe not for the right reason...

How do you check if props are equal? Deep or shallow?

Can I fool it by [using an array instead](https://github.com/mindplay-dk/fre/compare/fix-tests...fix-tests-2)?  (every value will be a new array instance)

(and how useful is this optimization? I haven't seen other libraries using this optimization - I suspect because most of the time, it doesn't make sense... as soon as you have an array or an event-handler in the props, this optimization won't work, right?)

Reply:
Props comparison is shallow, you can use array to cheat it~
To be honest, this is equivalent to that every component is `React.memo`, which is not effective most of the time.
This optimization is used in Vue [here](https://github.com/vuejs/vue-next/blob/e0a9cf5ace6e49d3c787ea9f78e741e0a4103070/packages/runtime-core/src/renderer.ts#L806)
It can reduce a part of rendering and prevent multiple initialization of hooks.

Reply:
> It can reduce a part of rendering and prevent multiple initialization of hooks.

In my opinion, this really isn't something you should trust the framework to do - it's too unpredictable. Someone adds a prop for an event-handler, for example, and suddenly the performance degrades with no obvious explanation.

I think `useMemo` is more predictable and transparent for optimizations like this.

Anyhow, I'll merge the updated test back to this branch now - it passes either way, but now it's passing for the right reason. 👍

Reply:
> it's too unpredictable. Someone adds a prop for an event-handler.

What is the performance degradation? If you pass events, it will lose optimization, but you can still use useMemo, as before.
We can open another issue to discuss this optimization, because before vue, few frameworks did this.

Reply:
> If you pass events, it will lose optimization, but you can still use useMemo, as before.

My point is, to understand the performance implications of adding an array or event-handler prop to an existing (unoptimized) codebase requires awareness and knowledge of implementation details.

To get predictable performance in day-to-day development, should the developer need to know how the engine is optimized internally?

As you pointed out, this detail affected a test in a way that was surprising to me - I'm not sure if it may be affecting some of the other tests too, but it might surprise anyone who's expecting to be able to generate some sort of side-effect. (not that we want to encourage side-effects, but not all developers know or care if side-effects are bad.)

Anyhow, that's just my personal point of view - I'm not opposed to optimizations, but I'm not fond of optimizations that only work under very specific conditions. Feel free to disagree. 😊

Reply:
> I'm not opposed to optimizations, but I'm not fond of optimizations that only work under very specific conditions.

Maybe it should be explained clearly in the document that when the shallow comparison results not equal, the optimization will be lost.

I have thought about this problem carefully. That built-in it can reduce the repeated initialization of hooks. The cost is to tell the developers the trigger condition of this optimization.